component/azstorage/block_blob.go

```diff
@@ -653,6 +655,200 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+	var blockOffset int64 = 0
+	blockList := common.BlockOffsetList{}
+	blobURL := bb.Container.NewBlockBlobURL(filepath.Join(bb.Config.prefixPath, name))
+	storageBlockList, err := blobURL.GetBlockList(
+		context.Background(), azblob.BlockListCommitted, bb.blobAccCond.LeaseAccessConditions)
+	if err != nil {
+		log.Err("BlockBlob::GetFileBlockOffsets : Failed to get block list %s ", name, err.Error())
+		return &common.BlockOffsetList{}, err
+	}
+	for _, block := range *&storageBlockList.CommittedBlocks {
+		blk := &common.Block{
+			Id:         block.Name,
+			StartIndex: int64(blockOffset),
+			EndIndex:   int64(blockOffset) + block.Size,
+			Size:       block.Size,
+		}
+		blockOffset += block.Size
+		blockList.BlockOffsetList = append(blockList.BlockOffsetList, blk)
+	}
+	if blockOffset == 0 {
+		blockList.SmallFile = true
+	}
+	return &blockList, nil
+}
+
+// create our definition of block
+func (bb *BlockBlob) createBlock(blockIdLength, startIndex, size int64) *common.Block {
+	newBlockId := base64.StdEncoding.EncodeToString(common.NewUUID(blockIdLength))
+	newBlock := &common.Block{
+		Id:         newBlockId,
+		StartIndex: startIndex,
+		EndIndex:   startIndex + size,
+		Size:       size,
+	}
+	return newBlock
+}
+
+func (bb *BlockBlob) createNewBlocks(modBlockList, blockList *common.BlockOffsetList, offset, length, blockIdLength int64) (bool, int64) {
+	prevIndex := blockList.BlockOffsetList[len(blockList.BlockOffsetList)-1].EndIndex
+	// BufferSize is the size of the buffer that will go beyond our current blob (appended)
+	var bufferSize int64
+	// counter will help us keep track of how many blocks we've created
+	counter := int64(0)
+	// appendOnly means there is no overlap at all with the blob - data only being appended
+	appendOnly := false
+	// if modBlockList is not of size 0 then we know that we're overwriting part of our current data - with new data also being appended to the end
+	if len(modBlockList.BlockOffsetList) != 0 {
+		bufferSize = length - (prevIndex - offset)
+		// only new data at the end basically offset => lastBlock.EndIndex
+	} else {
+		bufferSize = (offset + length) - prevIndex
+		appendOnly = true
+	}
+	// reset the offset to be the last index in the block list since we are adding new blocks
+	offset = prevIndex
+	startIndex := offset
+	for i := 0; i < int(bufferSize); i++ {
+		counter += 1
+		// create a new block if we hit our block size
+		if int64(i)%bb.Config.blockSize == 0 && i != 0 {
+			newBlock := bb.createBlock(blockIdLength, startIndex, bb.Config.blockSize)
+			modBlockList.BlockOffsetList = append(modBlockList.BlockOffsetList, newBlock)
+			blockList.BlockOffsetList = append(blockList.BlockOffsetList, newBlock)
+			startIndex = newBlock.EndIndex
+			// reset the counter since it will help us to determine if there is leftovers at the end
+			counter = 0
+		}
+	}
+	if counter != 0 {
+		newBlock := bb.createBlock(blockIdLength, startIndex, counter)
```

Comment:
These 3 lines are duplicate code, within loop also we are doing the same. unify them.